dali/operators/video/reader/video_reader_decoder_op.cc

```diff
@@ -0,0 +1,555 @@
+#include "dali/operators/video/video_utils.h"
+
+#include "libavutil/rational.h"
+
+namespace dali {
+
+struct VideoSampleDesc {
+  VideoSampleDesc(std::string filename = {}, int label = -1, int start = -1, int end = -1,
+                  int stride = -1)
+      : filename_(filename), label_(label), start_(start), end_(end), stride_(stride) {}
+  std::string filename_;
+  int label_ = -1;
+  int start_ = -1;
+  int end_ = -1;
+  int stride_ = -1;
+};
+
+template <typename Backend>
+struct VideoSample : public VideoSampleDesc {
+  VideoSample(std::string filename = {}, int label = -1, int start = -1, int end = -1,
+              int stride = -1)
+      : VideoSampleDesc{filename, label, start, end, stride} {}
+
+  VideoSample(const VideoSampleDesc &other) noexcept
+      : VideoSampleDesc(other) {}
+
+  // to be filled by Prefetch
+  Tensor<Backend> data_;
+  std::vector<double> timestamps_;
+  std::vector<int64_t> frame_idx_;
+};
+
+enum class FileListFormat {
+  kFrameIndex,
+  kTimestamp,
+  kTimestampInclusive
+};
+
+template <typename Backend, typename FramesDecoderImpl, typename Sample = VideoSample<Backend>>
+class VideoLoaderDecoder : public Loader<Backend, Sample, true> {
+ public:
+  explicit inline VideoLoaderDecoder(const OpSpec &spec)
+      : Loader<Backend, Sample, true>(spec),
+        file_root_(spec.GetArgument<std::string>("file_root")),
+        file_list_(spec.GetArgument<std::string>("file_list")),
+        filenames_(spec.GetRepeatedArgument<std::string>("filenames")),
+        sequence_len_(spec.GetArgument<int>("sequence_length")),
+        stride_(spec.GetArgument<int>("stride")),
+        step_(spec.GetArgument<int>("step")),
+        boundary_type_(GetBoundaryType(spec)) {
+    if ((spec.HasArgument("file_list") || spec.HasArgument("file_root") || spec.HasArgument("filenames")) != 1) {
+      DALI_FAIL("Only one of the following arguments can be provided: ``file_list``, ``file_root``, ``filenames``");
+    }
+    bool has_labels = spec.TryGetRepeatedArgument(labels_, "labels");
+    if (has_labels) {
+      DALI_ENFORCE(
+          labels_.size() == filenames_.size(),
+          make_string(
+              "Number of provided files and labels should match. Provided ",
+              filenames_.size(), " files and ", labels_.size(), " labels."));
+    }
+
+    video_files_info_ = GetVideoFiles(file_root_, filenames_, has_labels, labels_, file_list_);
+    DALI_ENFORCE(!video_files_info_.empty(), "No files were read.");
+
+    if (!file_list_.empty()) {
+      auto file_list_format_str = spec.GetArgument<std::string>("file_list_format");
+      if (file_list_format_str == "frame_index") {
+        file_list_format_ = FileListFormat::kFrameIndex;
+      } else if (file_list_format_str == "timestamp") {
+        file_list_format_ = FileListFormat::kTimestamp;
+      } else if (file_list_format_str == "timestamp_inclusive") {
+        file_list_format_ = FileListFormat::kTimestampInclusive;
+      }
```

Comment:
Is it ok to leave `file_list_format_` unset if `file_list_format_str` was something completely different? Maybe we should raise an error here?

Reply:
The code changed since this comment. We now have DALI_FAIL on the equivalent logic.